File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;
    const dx = [-1, 1, 0, 0];
    const dy = [0, 0, -1, 1];

   function bfs(sx, sy) {
    let visited = Array.from({ length: n }, () => Array(m).fill(-1));
    let queue = [];
    queue.push([sx, sy]);
    visited[sx][sy] = 1; // 시작 지점 거리 1

    while (queue.length > 0) {
        const [x, y] = queue.shift();

        for (let dir = 0; dir < 4; dir++) {
            const nx = x + dx[dir];
            const ny = y + dy[dir];

            if (
                nx >= 0 && nx < n &&
                ny >= 0 && ny < m &&
                maps[nx][ny] === 1 &&
                visited[nx][ny] === -1 // 방문 안 한 곳
            ) {
                visited[nx][ny] = visited[x][y] + 1;
                queue.push([nx, ny]);
            }
        }
    }

    return visited[n - 1][m - 1];
}


    return bfs(0, 0);
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Item{
    int x;
    int y;
    int count;
    
    Item(int x, int y, int count){
        this.x = x;
        this.y = y;
        this.count = count;
    }
}

class Solution {
    int[] dx = {-1,1,0,0};
    int[] dy = {0,0,-1,1};
    public int solution(int[][] maps) {
        int result = bfs(maps);
        return result;
    }
    
    int bfs(int[][] maps){
        int n = maps.length;
        int m = maps[0].length;
        
        Deque<Item> queue = new ArrayDeque<>();
        queue.offerFirst(new Item(0,0,1));
        boolean[][] visited = new boolean[n][m];
        visited[0][0] = true;
        
        while (!queue.isEmpty()){
            Item curr = queue.pollFirst();
            int x = curr.x;
            int y = curr.y;
            int count = curr.count;
            
            if (x == n - 1 && y == m - 1){
                return count;
            }
            
            for (int d = 0; d < 4; d++){
                int nx = x + dx[d];
                int ny = y + dy[d];
                
                if (0 <= nx && nx < n && 0 <= ny && ny < m && maps[nx][ny] == 1 && !visited[nx][ny]){
                    visited[nx][ny] = true;
                    queue.offerLast(new Item(nx,ny,count + 1));
                }
            }
        }
        return -1;
    }
}